### References

<br>

*__If you have used `fastqc-py` module, please cite:__*

*fastqc*: Andrews, Simon. "FastQC: a quality control tool for high throughput sequence data. 2010." (2017).

*multiqc*: Ewels, Philip, et al. "MultiQC: summarize analysis results for multiple tools and samples in a single report." Bioinformatics 32.19 (2016): 3047-3048.

*NanoPlot*: De Coster, Wouter, et al. "NanoPack: visualizing and processing long-read sequencing data." Bioinformatics 34.15 (2018): 2666-2669.

<br>

*__If you have used `filter_fastq-py` module, please cite:__*

*Porechop*: "Adapter trimmer for Oxford Nanopore reads." [https://github.com/rrwick/Porechop](https://github.com/rrwick/Porechop)

*Nanofilt*: De Coster, Wouter, et al. "NanoPack: visualizing and processing long-read sequencing data." Bioinformatics 34.15 (2018): 2666-2669.

<br>

*__If you have used `tax_assign-py` module, please cite:__*

*kraken2*: Wood, Derrick E., Jennifer Lu, and Ben Langmead. "Improved metagenomic analysis with Kraken 2." Genome biology 20.1 (2019): 1-13.

<br>

*__If you have used `tax_div-py` module, please cite:__*

*R programming language*: R Core Team. 2020. "R: A Language and Environment for Statistical Computing." Vienna, Austria: R Foundation for Statistical Computing. [https://www.R-project.org/]([https://www.R-project.org/).

*phyloseq*: McMurdie, P. J., & Holmes, S. (2013). "phyloseq: an R package for reproducible interactive analysis and graphics of microbiome census data." PloS one 8(4): e61217.
